DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
+    "        \n",
+    "    # Free all allocated resources of the session\n",
+    "    session.close()\n",
+    "    \n",
+    "    return hibernation_path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's introduce the main routine for training and inference. \n",
+    "It takes a started execution context (given by `execution_context` and `session`), an optional `epoch` index and a flag wheather to calculate WER report data (`query_report`) or not.\n",
+    "Its main duty is to iterate over all batches and calculate the mean loss.\n",
+    "If an epoch is provided, it will also do back propagation - aka training.\n",
+    "It will return a batch set result tuple which consists of the mean loss and - if asked for report data - the mean accuracy and individual results tuple."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def apply_batches(execution_context, session, epoch=-1, query_report=False):\n",
+    "    \n",
+    "    # An epoch of -1 means no train run\n",
+    "    do_training = epoch >= 0\n",
+    "    \n",
+    "    # Unpack variables\n",
+    "    if do_training:\n",
+    "        graph, data_set, tower_results, saver, apply_gradient_op, merged, writer = execution_context\n",
+    "    else:\n",
+    "        graph, data_set, tower_results, saver = execution_context\n",
+    "    results_params, _, avg_accuracy, avg_loss = tower_results\n",
+    "    \n",
+    "    batches_per_device = ceil(float(data_set.total_batches) / len(available_devices))\n",
     "\n",
-    "    # Apply gradients to modify the model\n",
-    "    apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)\n",
+    "    total_loss = 0.0\n",
+    "    params = [avg_loss]\n",
+    "    \n",
+    "    # Facilitate a train run\n",
+    "    if do_training:\n",
+    "        params.append([apply_gradient_op, merged])\n",
+    "    \n",
+    "    # Requirements to display a WER report\n",
+    "    if query_report:\n",
+    "        # Reset accuracy\n",
+    "        total_accuracy = 0.0\n",
+    "        # Create report results tuple\n",
+    "        report_results = ([],[],[],[])\n",
+    "        # Extend the session.run parameters\n",
+    "        params.append([results_params, avg_accuracy])\n",
     "\n",
-    "    # Create a saver to checkpoint the model\n",
-    "    saver = tf.train.Saver(tf.all_variables())\n",
+    "    # Loop over the batches\n",
+    "    for batch in range(int(batches_per_device)):\n",
+    "        extra_params = { }\n",
+    "        if do_training and do_fulltrace:\n",
+    "            loss_run_metadata            = tf.RunMetadata()\n",
+    "            extra_params['options']      = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)\n",
+    "            extra_params['run_metadata'] = loss_run_metadata\n",
```

Comment:
Rather than doing this whole block for every iteration, couldn't `extra_params` and `extra_params['options']` be initialised outside of the loop, and take out a bit of work here?